Infinite recursion when assigning to a variable of evolving type inside an array

<!-- BUGS: Please use this template. -->
<!-- QUESTIONS: This is not a general support forum! Ask Qs at http://stackoverflow.com/questions/tagged/typescript -->
<!-- SUGGESTIONS: See https://github.com/Microsoft/TypeScript-wiki/blob/master/Writing-Good-Design-Proposals.md -->

**TypeScript Version:**  2.0.0 / nightly

**Code**

```ts
I am getting the below error while integrating the plotly.js(v1.21.2) with typescript(2.0.0):

C:\Users\hemendra\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:36920
                throw e;
                ^
TypeError: Cannot read property 'id' of undefined
    at getNodeId (C:\Users\hemendra\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:13014:18)
    at getNodeLinks (C:\Users\hemendra\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:13402:26)
    at checkExpressionCached (C:\Users\hemendra\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:22848:25)
    at C:\Users\hemendra\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:15426:94
    at Object.map (C:\Users\hemendra\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:199:29)
    at getTypeOfVariableOrParameterOrProperty (C:\Users\hemendra\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:15426:44)
    at getTypeOfSymbol (C:\Users\hemendra\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:15561:24)
    at checkPropertyAccessExpressionOrQualifiedName (C:\Users\hemendra\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:21042:28)
    at checkPropertyAccessExpression (C:\Users\hemendra\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:21007:20)
    at checkExpressionWorker (C:\Users\hemendra\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:22937:28) 

Please find the attached zip containing the code,  Just extract the zip and run the below commands:
[TS_Plotly.zip](https://github.com/Microsoft/TypeScript/files/654959/TS_Plotly.zip)

npm i
tsc --sourceMap

While running tsc command i am getting error as specified above. Also i have tried with typescript@next but getting the same error.

```

**Expected behavior:**
It should integrate the plotly js and compile successful.

**Actual behavior:**
Getting error while including plotly.js.

@hemendray120 in the attached zip file plotly.js is missing. So I see below error:
```
error TS6053: File 'c:/repro/TS_Plotly/plotly.js' not found.
error TS6082: Only 'amd' and 'system' modules are supported alongside --out.
```

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -9469,9 +9469,17 @@ namespace ts {
         }
 
         function getAssignedTypeOfBinaryExpression(node: BinaryExpression): Type {
-            return node.parent.kind === SyntaxKind.ArrayLiteralExpression || node.parent.kind === SyntaxKind.PropertyAssignment ?
+            const isDestructuringDefaultAssignment =
+                node.parent.kind === SyntaxKind.ArrayLiteralExpression && isDestructuringAssignmentTarget(node.parent) ||
+                node.parent.kind === SyntaxKind.PropertyAssignment && isDestructuringAssignmentTarget(node.parent.parent);
+            return isDestructuringDefaultAssignment ?
                 getTypeWithDefault(getAssignedType(node), node.right) :
                 getTypeOfExpression(node.right);
+        }
+
+        function isDestructuringAssignmentTarget(parent: Node) {
+            return parent.parent.kind === SyntaxKind.BinaryExpression && (parent.parent as BinaryExpression).left === parent ||
+                parent.parent.kind === SyntaxKind.ForOfStatement && (parent.parent as ForOfStatement).initializer === parent;
         }
 
         function getAssignedTypeOfArrayLiteralElement(node: ArrayLiteralExpression, element: Expression): Type {

--- a/tests/baselines/reference/assignmentNestedInLiterals.js
+++ b/tests/baselines/reference/assignmentNestedInLiterals.js
@@ -5,9 +5,17 @@ target = [x = 1, y = x];
 var aegis, a, b;
 aegis = { x: a = 1, y: b = a };
 
+var kowloona, c, d;
+for (kowloona of [c = 1, d = c]) {
+}
+
 
 //// [assignmentNestedInLiterals.js]
 var target, x, y;
 target = [x = 1, y = x];
 var aegis, a, b;
 aegis = { x: a = 1, y: b = a };
+var kowloona, c, d;
+for (var _i = 0, _a = [c = 1, d = c]; _i < _a.length; _i++) {
+    kowloona = _a[_i];
+}

--- a/tests/baselines/reference/assignmentNestedInLiterals.js
+++ b/tests/baselines/reference/assignmentNestedInLiterals.js
@@ -1,0 +1,13 @@
+//// [assignmentNestedInLiterals.ts]
+var target, x, y;
+target = [x = 1, y = x];
+
+var aegis, a, b;
+aegis = { x: a = 1, y: b = a };
+
+
+//// [assignmentNestedInLiterals.js]
+var target, x, y;
+target = [x = 1, y = x];
+var aegis, a, b;
+aegis = { x: a = 1, y: b = a };

--- a/tests/baselines/reference/assignmentNestedInLiterals.symbols
+++ b/tests/baselines/reference/assignmentNestedInLiterals.symbols
@@ -1,0 +1,25 @@
+=== tests/cases/compiler/assignmentNestedInLiterals.ts ===
+var target, x, y;
+>target : Symbol(target, Decl(assignmentNestedInLiterals.ts, 0, 3))
+>x : Symbol(x, Decl(assignmentNestedInLiterals.ts, 0, 11))
+>y : Symbol(y, Decl(assignmentNestedInLiterals.ts, 0, 14))
+
+target = [x = 1, y = x];
+>target : Symbol(target, Decl(assignmentNestedInLiterals.ts, 0, 3))
+>x : Symbol(x, Decl(assignmentNestedInLiterals.ts, 0, 11))
+>y : Symbol(y, Decl(assignmentNestedInLiterals.ts, 0, 14))
+>x : Symbol(x, Decl(assignmentNestedInLiterals.ts, 0, 11))
+
+var aegis, a, b;
+>aegis : Symbol(aegis, Decl(assignmentNestedInLiterals.ts, 3, 3))
+>a : Symbol(a, Decl(assignmentNestedInLiterals.ts, 3, 10))
+>b : Symbol(b, Decl(assignmentNestedInLiterals.ts, 3, 13))
+
+aegis = { x: a = 1, y: b = a };
+>aegis : Symbol(aegis, Decl(assignmentNestedInLiterals.ts, 3, 3))
+>x : Symbol(x, Decl(assignmentNestedInLiterals.ts, 4, 9))
+>a : Symbol(a, Decl(assignmentNestedInLiterals.ts, 3, 10))
+>y : Symbol(y, Decl(assignmentNestedInLiterals.ts, 4, 19))
+>b : Symbol(b, Decl(assignmentNestedInLiterals.ts, 3, 13))
+>a : Symbol(a, Decl(assignmentNestedInLiterals.ts, 3, 10))
+

--- a/tests/baselines/reference/assignmentNestedInLiterals.symbols
+++ b/tests/baselines/reference/assignmentNestedInLiterals.symbols
@@ -23,3 +23,15 @@ aegis = { x: a = 1, y: b = a };
 >b : Symbol(b, Decl(assignmentNestedInLiterals.ts, 3, 13))
 >a : Symbol(a, Decl(assignmentNestedInLiterals.ts, 3, 10))
 
+var kowloona, c, d;
+>kowloona : Symbol(kowloona, Decl(assignmentNestedInLiterals.ts, 6, 3))
+>c : Symbol(c, Decl(assignmentNestedInLiterals.ts, 6, 13))
+>d : Symbol(d, Decl(assignmentNestedInLiterals.ts, 6, 16))
+
+for (kowloona of [c = 1, d = c]) {
+>kowloona : Symbol(kowloona, Decl(assignmentNestedInLiterals.ts, 6, 3))
+>c : Symbol(c, Decl(assignmentNestedInLiterals.ts, 6, 13))
+>d : Symbol(d, Decl(assignmentNestedInLiterals.ts, 6, 16))
+>c : Symbol(c, Decl(assignmentNestedInLiterals.ts, 6, 13))
+}
+

--- a/tests/baselines/reference/assignmentNestedInLiterals.types
+++ b/tests/baselines/reference/assignmentNestedInLiterals.types
@@ -1,0 +1,35 @@
+=== tests/cases/compiler/assignmentNestedInLiterals.ts ===
+var target, x, y;
+>target : any
+>x : any
+>y : any
+
+target = [x = 1, y = x];
+>target = [x = 1, y = x] : number[]
+>target : any
+>[x = 1, y = x] : number[]
+>x = 1 : 1
+>x : any
+>1 : 1
+>y = x : number
+>y : any
+>x : number
+
+var aegis, a, b;
+>aegis : any
+>a : any
+>b : any
+
+aegis = { x: a = 1, y: b = a };
+>aegis = { x: a = 1, y: b = a } : { x: number; y: number; }
+>aegis : any
+>{ x: a = 1, y: b = a } : { x: number; y: number; }
+>x : number
+>a = 1 : 1
+>a : any
+>1 : 1
+>y : number
+>b = a : number
+>b : any
+>a : number
+

--- a/tests/baselines/reference/assignmentNestedInLiterals.types
+++ b/tests/baselines/reference/assignmentNestedInLiterals.types
@@ -33,3 +33,19 @@ aegis = { x: a = 1, y: b = a };
 >b : any
 >a : number
 
+var kowloona, c, d;
+>kowloona : any
+>c : any
+>d : any
+
+for (kowloona of [c = 1, d = c]) {
+>kowloona : any
+>[c = 1, d = c] : number[]
+>c = 1 : 1
+>c : any
+>1 : 1
+>d = c : number
+>d : any
+>c : number
+}
+

--- a/tests/cases/compiler/assignmentNestedInLiterals.ts
+++ b/tests/cases/compiler/assignmentNestedInLiterals.ts
@@ -4,3 +4,7 @@ target = [x = 1, y = x];
 
 var aegis, a, b;
 aegis = { x: a = 1, y: b = a };
+
+var kowloona, c, d;
+for (kowloona of [c = 1, d = c]) {
+}

--- a/tests/cases/compiler/assignmentNestedInLiterals.ts
+++ b/tests/cases/compiler/assignmentNestedInLiterals.ts
@@ -1,0 +1,6 @@
+// @noImplicitAny: true
+var target, x, y;
+target = [x = 1, y = x];
+
+var aegis, a, b;
+aegis = { x: a = 1, y: b = a };

